enumerable.rb

```diff
@@ -0,0 +1,130 @@
+module Enumerable
+  def my_each
+    return to_enum(my_each) unless block_given?
+
+    i = 0
+    while i < size
+      yield(to_a[i])
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    rreturn to_enum(my_each_with_index) unless block_given?
+
+    i = 0
+    while i < size
+      yield(to_a[i], i)
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return to_enum(my_select) unless block_given?
+
+    result = []
+    my_each do |item|
+      result << item if yield(item)
+    end
+    result
+  end
+
+  def my_all?(param = nil)
+    result = true
+    my_each do |value|
+      if block_given?
+        result = false unless yield(value)
+      elsif param.nil?
+        result = true unless value
+      else
+        result = false unless param === value
+      end
+    end
+    result
+  end
+
+  def my_any?(param = nil)
+    result = false
+    my_each do |value|
+      if block_given?
+        result = true if yield(value)
+      elsif param.nil?
+        result = true if value
+      elsif param === value
+        result = true
+      end
+    end
+    result
+  end
+
+  def my_none?(param = nil)
+    result = true
+    my_each do |value|
+      if block_given?
+        result = false if yield(value)
+      elsif param.nil?
+        result = false if value
+      elsif param === value
+        result = false
+      end
+    end
+    result
+  end
+
+  def my_count(param = nil)
+    count = 0
+    my_each do |item|
+      if param
+        count += 1 if item == param
+      elsif block_given?
+        count += 1 if yield(item)
+      else
+        count = length
+      end
+    end
+    count
+  end
+
+  def my_map(param = nil)
+    return to_enum(__method__) unless block_given?
+
+    arr = []
+    my_each do |item|
+      if param
+        a = param.call(item)
+        return arr << a
+      else
+        arr << yield(item)
+      end
+    end
+    arr
+  end
```

Comment:
## [OPTIONAL]

#### my_map 

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly advice working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

Reply:
done